classes/local/form/login_form.php

```diff
@@ -33,24 +33,40 @@ class login_form extends \moodleform
      * {@inheritDoc}
      * @see moodleform::definition()
      */
-    public function definition()
-    {
-        global $OUTPUT;
+    public function definition() {
         $mform = $this->_form;
+        $factorname = $this->_customdata['factor_name'];
+
+        $mform->addElement('hidden', 'factor_name', $factorname);
+        $mform->setType('factor_name', PARAM_ALPHA);
 
-        $factors = \tool_mfa\plugininfo\factor::get_enabled_factors();
-        foreach ($factors as $factor) {
+        if ($factorname == 'grace') {
+            $mform = $this->define_grace_period_page($mform);
+        } else {
+            $factor = \tool_mfa\plugininfo\factor::get_factor($factorname);
             $mform = $factor->define_login_form($mform);
+            $this->add_action_buttons();
         }
+    }
+
+    public function define_grace_period_page($mform) {
+        global $OUTPUT;
+
+        $mform->addElement('html', $OUTPUT->heading(get_string('graceperiod:notconfigured', 'tool_mfa'), 3));
+        $mform->addElement('html', $OUTPUT->heading(get_string('graceperiod:canaccess', 'tool_mfa'), 5));
+
+        // TODO: get grace period expiration date
+        $mform->addElement('html', $OUTPUT->heading(get_string('graceperiod:expires', 'tool_mfa', time()), 5));
+        $this->add_action_buttons(false, get_string('ok'));
 
-        $this->add_action_buttons();
+        return $mform;
     }
 
     public function validation($data, $files) {
         $errors = parent::validation($data, $files);
 
-        $factors = \tool_mfa\plugininfo\factor::get_enabled_factors();
-        foreach ($factors as $factor) {
+        $factor = \tool_mfa\plugininfo\factor::get_factor($data['factor_name']);
+        if ($factor) {
```

Comment:
if $factor is not valid lets make that an $error too

Reply:
I re-wrote this block and implemented new parameter `grace_mode`, so now there is no need to check `if ($factor)`